File: flightnotify/domain/fingerprints.py
"""Stable hashes for provider queries, itineraries, configs and alerts.

Every fingerprint is a SHA-256 hex digest over a canonical JSON encoding, so
the same logical input always produces the same key across processes and
restarts.
"""

from __future__ import annotations

import hashlib
import json
from datetime import date
from decimal import Decimal
from typing import Any


def _canonical(value: Any) -> Any:
    if isinstance(value, Decimal):
        # Quantize so 100 and 100.00 hash identically.
        return format(value.normalize(), "f")
    if isinstance(value, date):
        return value.isoformat()
    if isinstance(value, dict):
        return {str(k): _canonical(v) for k, v in sorted(value.items()) if v is not None}
    if isinstance(value, list | tuple):
        return [_canonical(v) for v in value]
    if isinstance(value, bool | int | float | str) or value is None:
        return value
    return str(value)


def digest(payload: dict[str, Any]) -> str:
    blob = json.dumps(_canonical(payload), sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(blob.encode("utf-8")).hexdigest()


def query_fingerprint(endpoint: str, params: dict[str, Any]) -> str:
    """Identify a provider request. The API key is never part of the key."""
    scrubbed = {k: v for k, v in params.items() if k not in {"api_key", "output", "no_cache"}}
    return digest({"endpoint": endpoint, "params": scrubbed})


def itinerary_fingerprint(
    *,
    origin: str | None,
    destination: str | None,
    outbound_date: date | None,
    return_date: date | None,
    flight_numbers: list[str] | None,
    departure_time: str | None,
    arrival_time: str | None,
    stops: int | None,
    market: str,
) -> str:
    """Identify a specific itinerary so repeats can be deduplicated."""
    return digest(
        {
            "origin": origin,
            "destination": destination,
            "outbound_date": outbound_date,
            "return_date": return_date,
            "flight_numbers": list(flight_numbers or []),
            "departure_time": departure_time,
            "arrival_time": arrival_time,
            "stops": stops,
            "market": market,
        }
    )


def config_fingerprint(payload: dict[str, Any]) -> str:
    """Identify a comparison series (comparison-relevant settings only)."""
    return digest(payload)


def alert_dedupe_key(
    *,
    tracker_id: int,
    config_version_id: int | None,
    alert_type: str,
    price: Decimal,
    currency: str,
    itinerary_fingerprint_value: str | None,
    outbound_date: date | None,
    return_date: date | None,
    market: str | None,
) -> str:
    """A repeat of the same finding must map to the same key."""
    return digest(
        {
            "tracker_id": tracker_id,
            "config_version_id": config_version_id,
            "alert_type": alert_type,
            "price": price.quantize(Decimal("0.01")),
            "currency": currency,
            "itinerary": itinerary_fingerprint_value,
            "outbound_date": outbound_date,
            "return_date": return_date,
            "market": market,
        }
    )

File: worker/src/domain/fingerprints.ts
/**
 * Stable hashes for provider queries, itineraries, configs and alerts.
 *
 * Port of `flightnotify/domain/fingerprints.py`. Every digest is SHA-256 over
 * the canonical JSON encoding, so the same logical input produces the same key
 * across processes, restarts -- and now across the Python/TypeScript boundary,
 * which is what lets the existing production rows keep their meaning after
 * import.
 *
 * These are async because WebCrypto is; the Python originals were sync.
 */

import { canonicalJson, type CanonicalInput } from "./canonical.js";
import { decimalStringFromCents } from "./money.js";

const encoder = new TextEncoder();

export async function sha256Hex(text: string): Promise<string> {
  const buffer = await crypto.subtle.digest("SHA-256", encoder.encode(text));
  const bytes = new Uint8Array(buffer);
  let out = "";
  for (const byte of bytes) out += byte.toString(16).padStart(2, "0");
  return out;
}

export async function digest(payload: CanonicalInput): Promise<string> {
  return sha256Hex(canonicalJson(payload));
}

/** Parameters that must never enter a cache/---fingerprint key. */
const EXCLUDED_QUERY_PARAMS = new Set(["api_key", "output", "no_cache"]);

/** Identify a provider request. The API key is never part of the key. */
export async function queryFingerprint(
  endpoint: string,
  params: Record<string, CanonicalInput>,
): Promise<string> {
  const scrubbed: Record<string, CanonicalInput> = {};
  for (const [key, value] of Object.entries(params)) {
    if (!EXCLUDED_QUERY_PARAMS.has(key)) scrubbed[key] = value;
  }
  return digest({ endpoint, params: scrubbed });
}

export interface ItineraryFingerprintArgs {
  origin: string | null;
  destination: string | null;
  outbound_date: string | null;
  return_date: string | null;
  flight_numbers: string[] | null;
  departure_time: string | null;
  arrival_time: string | null;
  stops: number | null;
  market: string;
}

/** Identify a specific itinerary so repeats can be deduplicated. */
export async function itineraryFingerprint(
  args: ItineraryFingerprintArgs,
): Promise<string> {
  return digest({
    origin: args.origin,
    destination: args.destination,
    outbound_date: args.outbound_date,
    return_date: args.return_date,
    // Python coerces a missing list to [], which survives canonicalisation
    // (an empty list is not None), so it must not become a dropped key here.
    flight_numbers: args.flight_numbers ?? [],
    departure_time: args.departure_time,
    arrival_time: args.arrival_time,
    stops: args.stops,
    market: args.market,
  });
}

/** Identify a comparison series (comparison-relevant settings only). */
export async function configFingerprint(
  payload: Record<string, CanonicalInput>,
): Promise<string> {
  return digest(payload);
}

export interface AlertDedupeArgs {
  tracker_id: number;
  config_version_id: number | null;
  alert_type: string;
  price_cents: number;
  currency: string;
  itinerary_fingerprint_value: string | null;
  outbound_date: string | null;
  return_date: string | null;
  market: string | null;
}

/**
 * A repeat of the same finding must map to the same key.
 *
 * The price is rendered through `decimalStringFromCents`, which reproduces
 * Python's `format(Decimal.normalize(), "f")`. That is why $1042 and $1042.00
 * collapse to one key instead of alerting twice for the same fare.
 */
export async function alertDedupeKey(args: AlertDedupeArgs): Promise<string> {
  return digest({
    tracker_id: args.tracker_id,
    config_version_id: args.config_version_id,
    alert_type: args.alert_type,
    price: decimalStringFromCents(args.price_cents),
    currency: args.currency,
    itinerary: args.itinerary_fingerprint_value,
    outbound_date: args.outbound_date,
    return_date: args.return_date,
    market: args.market,
  });
}
